src/logic/algo/solver.ts

```diff
@@ -23,20 +23,124 @@ export type StepToImplementName = keyof Solver;
 export class MonteCarloTreeSearch implements Solver {
   private kernel: NonNullable<MonteCarloTreeSearchHyperParams["kernel"]> = ubcKernel;
 
+  private nodeRewards: Map<TreeNode, number> = new Map();
+
+  private nodeVisits: Map<TreeNode, number> = new Map();
+
+  private nodeChildren: Map<TreeNode, Array<TreeNode>> = new Map();
+
   run(root: TreeNode, hyperParams: MonteCarloTreeSearchHyperParams): TreeNode {
+    while (true) {
+      for (let i = 0; i < hyperParams.numIterations; i += 1) {
+        this.singleRun(root, hyperParams);
+      }
+      root = this.choose(root);
+      if (root.children.length === 0) {
+        return root;
+      }
+    }
+  }
+
+  singleRun(root: TreeNode, hyperParams: MonteCarloTreeSearchHyperParams): void {
     this.kernel = hyperParams.kernel ?? this.kernel;
-    return new TreeNode();
+    const path = this.select(root);
+    const leaf = path[path.length - 1];
+    this.expand(leaf);
+    let reward = 0;
+    for (let i = 0; i < hyperParams.numRollout; i += 1) {
+      reward += this.simulate(leaf);
+    }
+    this.backup(path, reward);
   }
 
-  choose(node: TreeNode): null {
-    return null;
+  select(node: TreeNode): Array<TreeNode> {
+    const path: Array<TreeNode> = [];
+    while (true) {
+      path.push(node);
+      if (!this.nodeChildren.has(node)) {
+        return path;
+      }
+      if (this.nodeChildren.get(node) ?? [].length === 0) {
```

Comment:
Można też `if (this.nodeChildren.get(node) ?? true)`
I też pytanie: jeśli my sprawdzamy czy node istnieje to czy to już nie jest zrobione wyżej
![image](https://user-images.githubusercontent.com/21079319/116480885-7eb5f180-a882-11eb-8833-967087a40291.png)

Reply:
Tutaj chciałam sprawdzić, czy node istnieje, ale nie ma przypisanych dzieci

Reply:
To w takim razie chyba brakuje ci nawiasu, ponieważ teraz to wygląda tak, że po tych dwóch pytajnikach będzie zwrócona lista dzieci albo true (`[].length === 0`)
![image](https://user-images.githubusercontent.com/21079319/116569327-93899800-a909-11eb-902e-82ece339245f.png)


PS. 
void 0 - to shortcut do undefined'a xd
![image](https://user-images.githubusercontent.com/21079319/116569454-b025d000-a909-11eb-9df7-f57c545be4c6.png)

PS. PS
TS Playground - to online kompilator ts'a